bazel/protobuf.bzl

```diff
@@ -165,6 +179,64 @@ java_proto_gen_validate = rule(
             executable = True,
         ),
     },
-    output_to_genfiles = True,
-    implementation = _protoc_gen_validate_java_impl,
+)
+
+def _java_proto_gen_validate_impl(ctx):
+    source_jars = [source_jar for dep in ctx.attr.deps for source_jar in dep[_ProtoValidateSourceInfo].sources]
+
+    deps = [java_common.make_non_strict(dep[JavaInfo]) for dep in ctx.attr.java_deps]
+    deps += [dep[JavaInfo] for dep in ctx.attr._validate_deps]
+
+    java_info = java_common.compile(
+        ctx,
+        source_jars = source_jars,
+        deps = deps,
+        output_source_jar = ctx.outputs.srcjar,
+        output = ctx.outputs.jar,
+        java_toolchain = find_java_toolchain(ctx, ctx.attr._java_toolchain),
+        host_javabase = find_java_runtime_toolchain(ctx, ctx.attr._host_javabase),
+    )
+
+    return [java_info]
+
+"""Bazel rule to create a Java protobuf validation library from proto sources files.
+
+Args:
+  deps: proto_library rules that contain the necessary .proto files
+  java_deps: the java_proto_library of the protos being compiled.
```

Comment:
I would love to remove the redundancy, I just don't see a good way to do it.

1. The rule needs a proto source as argument be able to feed it to protoc. I do not know of a way to retrieve that source from a java_proto_library target.

1. Definitely doable, not sure if it's desirable. This would mean that the java proto generated code in a project is either generated by the validate rules or generated by both the validate and the native java_proto_library rules (if both are specified). Having duplicates of the same class could be a problem. It would also mean that the validation target will become a dependency for other rules that need the protobuf code (such as grpc). Lastly, the grpc-java purposely does not generate this code, although I'm unsure as to why exactly (grpc/grpc-java#5237 might provide some context).

I'll leave the decision up to you.

Reply:
Thanks for the gRPC link, that was helpful. Let's leave this as-is, then.